lib/internal/test_runner/mock.js

```diff
@@ -0,0 +1,303 @@
+'use strict';
+const {
+  ArrayPrototypePush,
+  ArrayPrototypeSlice,
+  Error,
+  ObjectDefineProperty,
+  ObjectGetOwnPropertyDescriptor,
+  Proxy,
+  ReflectApply,
+  ReflectConstruct,
+  ReflectGet,
+  SafeMap,
+} = primordials;
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+    ERR_INVALID_ARG_VALUE,
+  }
+} = require('internal/errors');
+const { kEmptyObject } = require('internal/util');
+const {
+  validateBoolean,
+  validateFunction,
+  validateInteger,
+  validateObject,
+} = require('internal/validators');
+
+function kDefaultFunction() {}
+
+class MockFunctionContext {
+  #calls;
+  #mocks;
+  #implementation;
+  #restore;
+  #times;
+
+  constructor(implementation, restore, times) {
+    this.#calls = [];
+    this.#mocks = new SafeMap();
+    this.#implementation = implementation;
+    this.#restore = restore;
+    this.#times = times;
+  }
+
+  get calls() {
+    return ArrayPrototypeSlice(this.#calls, 0);
+  }
+
+  callCount() {
+    return this.#calls.length;
+  }
+
+  mockImplementation(implementation) {
+    validateFunction(implementation, 'implementation');
+    this.#implementation = implementation;
+  }
+
+  mockImplementationOnce(implementation, onCall) {
+    validateFunction(implementation, 'implementation');
+    const nextCall = this.#calls.length;
+    const call = onCall ?? nextCall;
+    validateInteger(call, 'onCall', 0);
+
+    if (call < nextCall) {
+      // The call number has already passed.
+      return;
+    }
+
+    this.#mocks.set(call, implementation);
+  }
+
+  restore() {
+    const { descriptor, object, original, methodName } = this.#restore;
+
+    if (typeof methodName === 'string') {
+      // This is an object method spy.
+      ObjectDefineProperty(object, methodName, descriptor);
+    } else {
+      // This is a bare function spy. There isn't much to do here but make
+      // the mock call the original function.
+      this.#implementation = original;
+    }
+  }
+
+  trackCall(call) {
+    ArrayPrototypePush(this.#calls, call);
+  }
+
+  nextImpl() {
+    const nextCall = this.#calls.length;
+    const mock = this.#mocks.get(nextCall);
+    const impl = mock ?? this.#implementation;
+
+    if (nextCall + 1 === this.#times) {
+      this.restore();
+    }
+
+    this.#mocks.delete(nextCall);
+    return impl;
+  }
+}
+
+const { nextImpl, restore, trackCall } = MockFunctionContext.prototype;
+delete MockFunctionContext.prototype.trackCall;
+delete MockFunctionContext.prototype.nextImpl;
+
+class MockTracker {
+  #mocks;
+
+  constructor() {
+    this.#mocks = [];
+  }
+
+  fn(
+    original = function() {},
```

Comment:
```suggestion
    original = kDefaultFunction,
```

Reply:
Using `function() {}` was deliberate here and on line 121 so that if someone writes through the proxy and adds properties to the function, those won't appear on future mocks that reuse the same noop function (there is a test that verifies this).